In [1]:
import numpy as np
import pandas as pd
from sklearn.metrics import dcg_score,ndcg_score
from collections import OrderedDict

# **Welcome to week 4 project!**

Welcome to the last week of the course -- so excited to see that you've made it to the end! 👏 

We’ve already discussed the importance of measuring model performance. As Lord Kelvin said, “To measure is to know – If you cannot measure it, you cannot improve it.” And he was right – metrics are the only way we can actually evaluate our model’s performance!

In this week's project, we will touch upon two key aspects related to evaluation:
1. Behavioral metrics
2. Off-policy evaluation

For behavioral metrics, we will work with Spotify music sessions dataset, and implement a few behavioral metrics and understand their relationships with traditional metrics.

For off-policy evaluation, we will simulate a dataset where we have logged action policies, and see how IPS is implemented.


### Goals for this week's project:

For this week's project assignment, we will complete the following tasks:
1. Implement 3 behavioral metrics and present the correlation plot for them
2. Implement another ranking logic (e.g. sort by track popularity, or sort by danceability and compare all metrics on productional ranking logic and this new ranking logic.
3. Complete the implementation of two off-policy estimators: Capped IPS and Normalized Capped Importance Sampling (NCIS)


# Part A: Behavioral metrics

Behavioral metrics include factors like what items a user interacts with and how, the amount of time they spend on the platform, and how they spend that time.

To define and implement a few behavioral metrics, we will work with the Spotify music sessions dataset.
Download the dataset from GDrive: https://drive.google.com/drive/folders/10LGZMgXRuz2qPr_QDbYdVVlKEcnQ25YL?usp=sharing
(files: log_mini.cvs and tf_mini.csv)

## Spotify music sessions dataset

The public part of the dataset consists of roughly 130 million listening sessions with associated user interactions on the Spotify service. In total, users interacted with almost 4 million tracks during these sessions, and the dataset includes acoustic features and metadata for all of these tracks.

Detailed description of the dataset can be found here:
https://drive.google.com/file/d/1BELTuH4nBeyHna5EAGzJv-HWHKrbxPsf/view?usp=sharing



In [2]:
log = pd.read_csv("../data/week4_data/log_mini.csv")
tracks = pd.read_csv("../data/week4_data/tf_mini.csv")

In [7]:
log.head(21)

session_id  session_position  session_length  \
0   0_00006f66-33e5-4de7-a324-2d18e439fc1e                 1              20   
1   0_00006f66-33e5-4de7-a324-2d18e439fc1e                 2              20   
2   0_00006f66-33e5-4de7-a324-2d18e439fc1e                 3              20   
3   0_00006f66-33e5-4de7-a324-2d18e439fc1e                 4              20   
4   0_00006f66-33e5-4de7-a324-2d18e439fc1e                 5              20   
5   0_00006f66-33e5-4de7-a324-2d18e439fc1e                 6              20   
6   0_00006f66-33e5-4de7-a324-2d18e439fc1e                 7              20   
7   0_00006f66-33e5-4de7-a324-2d18e439fc1e                 8              20   
8   0_00006f66-33e5-4de7-a324-2d18e439fc1e                 9              20   
9   0_00006f66-33e5-4de7-a324-2d18e439fc1e                10              20   
10  0_00006f66-33e5-4de7-a324-2d18e439fc1e                11              20   
11  0_00006f66-33e5-4de7-a324-2d18e439fc1e                12              20   
12  0_00006f66-33e5-4de7-a324-2d18e439fc1e                13              20   
13  0_00006f66-33e5-4de7-a324-2d18e439fc1e                14              20   
14  0_00006f66-33e5-4de7-a324-2d18e439fc1e                15              20   
15  0_00006f66-33e5-4de7-a324-2d18e439fc1e                16              20   
16  0_00006f66-33e5-4de7-a324-2d18e439fc1e                17              20   
17  0_00006f66-33e5-4de7-a324-2d18e439fc1e                18              20   
18  0_00006f66-33e5-4de7-a324-2d18e439fc1e                19              20   
19  0_00006f66-33e5-4de7-a324-2d18e439fc1e                20              20   
20  0_0000a72b-09ac-412f-b452-9b9e79bded8f                 1              20   

                            track_id_clean  skip_1  skip_2  skip_3  \
0   t_0479f24c-27d2-46d6-a00c-7ec928f2b539   False   False   False   
1   t_9099cd7b-c238-47b7-9381-f23f2c1d1043   False   False   False   
2   t_fc5df5ba-5396-49a7-8b29-35d0d28249e0   False   False   False   
3   t_23cff8d6-d874-4b20-83dc-94e450e8aa20   False   False   False   
4   t_64f3743c-f624-46bb-a579-0f3f9a07a123   False   False   False   
5   t_c815228b-3212-4f9e-9d4f-9cb19b248184   False   False    True   
6   t_e23c19f5-4c32-4557-aa44-81372c2e3705    True    True    True   
7   t_0be6eced-f56f-48bd-8086-f2e0b760fdee    True    True    True   
8   t_f3ecbd3b-9e8e-4557-b8e0-39cfcd7e65dd   False    True    True   
9   t_2af4dfa0-7df3-4b7e-b7ab-353ba48237f9    True    True    True   
10  t_79570b2a-639a-4ec0-9853-71c5299ac44d    True    True    True   
11  t_69016f19-84aa-40c0-afa9-54404397b7a2    True    True    True   
12  t_59dc3fcd-7aec-4da5-a747-b59b19bab3bb    True    True    True   
13  t_a157ba98-7eae-4e7d-99d5-ed760b2c0978   False   False   False   
14  t_87d95b75-af5c-4ef6-8dc4-cd888ae17cce   False    True    True   
15  t_db945033-3dc5-4a22-8889-c72fd2781299   False    True    True   
16  t_bff5b6c6-6968-41be-b723-7e860b372975   False    True    True   
17  t_bf496d77-9eb2-4fe1-80f7-2c7ecd6f9b8e   False    True    True   
18  t_1051bc37-1a73-4301-812b-f83ae0c9bbcd    True    True    True   
19  t_358c9cce-7a1e-4dd4-81de-206dda80363f    True    True    True   
20  t_e975c3a1-6a6c-4955-8bc7-02eb8e701bec   False    True    True   

    not_skipped  context_switch  no_pause_before_play  ...  \
0          True               0                     0  ...   
1          True               0                     1  ...   
2          True               0                     1  ...   
3          True               0                     1  ...   
4          True               0                     1  ...   
5         False               0                     1  ...   
6         False               0                     1  ...   
7         False               0                     1  ...   
8         False               0                     1  ...   
9         False               0                     1  ...   
10        False               0                 

In [5]:
tracks.head(2)

track_id    duration  release_year  \
0  t_a540e552-16d4-42f8-a185-232bd650ea7d  109.706673          1950   
1  t_67965da0-132b-4b1e-8a69-0ef99b32287c  187.693329          1950   

   us_popularity_estimate  acousticness  beat_strength  bounciness  \
0               99.975414      0.458040       0.519497    0.504949   
1               99.969430      0.916272       0.419223    0.545530   

   danceability  dyn_range_mean    energy  ...  time_signature   valence  \
0      0.399767        7.511880  0.817709  ...               4  0.935512   
1      0.491235        9.098376  0.154258  ...               3  0.359675   

   acoustic_vector_0  acoustic_vector_1  acoustic_vector_2  acoustic_vector_3  \
0          -0.033284          -0.411896           -0.02858           0.349438   
1           0.145703          -0.850372            0.12386           0.746904   

  acoustic_vector_4  acoustic_vector_5  acoustic_vector_6  acoustic_vector_7  
0          0.832467          -0.213871          -0.299464          -0.675907  
1          0.371803          -0.420558          -0.213120          -0.525795  

[2 rows x 30 columns]

### Traditional metrics: ndcg@k

We will use the log dataframe as the main dataframe for evaluation of metrics. The skip_1 flag can be used as a relevance signal -- if the user found the recommendation relevant, skip_1 = False. With this relevance signal, we can compute simple ndcg metrics -- one for each session and then averaged across all sessions. This will serve as a base metric for comparison.

Note: the ranking logic here is assumed to be the production ranker, i.e. sorting by session_position gives the exact order of tracks the Spotify ranker presented to the user.

Lets compute simple skip rate and ndcg metric for the production ranker:

In [8]:
topk = 10

skip_rate_session = log.groupby('session_id').apply\
(lambda x: x.nsmallest(topk,['session_position'])).reset_index(drop = True).groupby("session_id").skip_1.mean().mean()
print("average skip rate in the session: ",skip_rate_session)

average skip rate in the session:  0.40893999999999464


### Average skip rate in top-10 positions for each session and then average of that.

In [10]:
log.groupby('session_id').apply\
(lambda x: x.nsmallest(topk,['session_position'])).reset_index(drop = True).groupby("session_id").skip_1.mean()

session_id
0_00006f66-33e5-4de7-a324-2d18e439fc1e    0.3
0_0000a72b-09ac-412f-b452-9b9e79bded8f    0.6
0_00010fc5-b79e-4cdf-bc4c-f140d0f99a3a    0.5
0_00016a3d-9076-4f67-918f-f29e3ce160dc    0.7
0_00018b58-deb8-4f98-ac5e-d7e01b346130    0.1
                                         ... 
0_0eac164c-f209-4590-8608-a56e67658952    1.0
0_0eacbee7-9868-48a0-9ab0-f86069329f50    0.8
0_0ead11fc-f32c-4eb4-8fc1-15b51432a404    0.6
0_0eae096d-ae4c-4556-8227-0f6294958330    0.6
0_0eaeef5d-25e9-4429-bd55-af15d3604c9f    0.0
Name: skip_1, Length: 10000, dtype: float64

In [12]:
topk = 10

def get_ndcg(df):
    true_relevance = np.asarray(1-np.asarray(df['skip_1']*1.0))
    ranker_scores = np.asarray(1/np.asarray(df['session_position']))
    # approximate the ranker scores using the session position
    return(ndcg_score([true_relevance], [ranker_scores]))


ndcg_metric = log.groupby('session_id').apply\
(lambda x: x.nsmallest(topk,['session_position'])).reset_index(drop = True).groupby("session_id").apply(get_ndcg)\
.mean()
print("NDCG@k, with k= ",topk," is: ",ndcg_metric)

NDCG@k, with k=  10  is:  0.8330266041453124


## Goals for this week:

Implement a few behavioral metrics and compare their correlations. We will implement the following three metrics:
1. *Time to first skip:* how long did it take for the user to get the first bad recommendation, i.e. a recommendation they skipped. Since we can't easily calculate time, we can use number of songs as a proxy and compute the metric as number of songs it took for the first skip.

2. *Sustained dissatisfaction:* we assume that the user is dissatisfied in a sustained manner if they skip 3 songs consecutively.

3. *Session coherence:* we define coherence as how similar the recommended musical tracks are. We can use the acoustic_vector of the music tracks to calculate the similarity.

In [15]:
def time_to_skip(df, topk):
    count = 0
    for index, row in df.iterrows():
        if row['skip_1'] == True:
            return count
        count += 1
    return topk

In [17]:
log.groupby('session_id').apply\
(lambda x: x.nsmallest(topk,['session_position'])).reset_index(drop = True).groupby("session_id").\
apply(time_to_skip, topk).mean()

3.2057

### Find the maximum of consecutive True in column skip_1

In [30]:
def sustained_dissatisfaction(df):
    b = df['skip_1'].cumsum()
#     print(b)
    c = b.sub(b.mask(df['skip_1']).ffill().fillna(0)).astype(int)
#     print(c)
    mask = df['skip_1'].any()
    length = np.where(mask, c.max(), -1)
#     print(length)
    if length >= 3:
        return 1
    return 0

In [32]:
log.groupby('session_id').apply\
(lambda x: x.nsmallest(topk,['session_position'])).reset_index(drop = True).groupby("session_id").\
apply(sustained_dissatisfaction).mean()

0.5238

In [33]:
from sklearn.metrics.pairwise import cosine_similarity

In [34]:
log_tracks = pd.merge(log, tracks, left_on = 'track_id_clean', right_on = 'track_id', how = 'left')

In [35]:
log_tracks.head(2)

session_id  session_position  session_length  \
0  0_00006f66-33e5-4de7-a324-2d18e439fc1e                 1              20   
1  0_00006f66-33e5-4de7-a324-2d18e439fc1e                 2              20   

                           track_id_clean  skip_1  skip_2  skip_3  \
0  t_0479f24c-27d2-46d6-a00c-7ec928f2b539   False   False   False   
1  t_9099cd7b-c238-47b7-9381-f23f2c1d1043   False   False   False   

   not_skipped  context_switch  no_pause_before_play  ...  time_signature  \
0         True               0                     0  ...               4   
1         True               0                     1  ...               4   

    valence  acoustic_vector_0  acoustic_vector_1  acoustic_vector_2  \
0  0.152255          -0.815775           0.386409           0.230160   
1  0.337152          -0.713646           0.363718           0.310315   

   acoustic_vector_3 acoustic_vector_4  acoustic_vector_5 acoustic_vector_6  \
0           0.028028         -0.333373           0.015452         -0.353590   
1          -0.042222         -0.383164           0.066357         -0.365308   

  acoustic_vector_7  
0          0.205826  
1          0.157920  

[2 rows x 51 columns]

In [37]:
def similarity(df):
    df_subset = df[df.columns[df.columns.str.startswith('acoustic_vector')]]
    return cosine_similarity(df_subset).mean()

In [39]:
log_tracks.groupby('session_id').apply\
(lambda x: x.nsmallest(topk,['session_position'])).reset_index(drop = True).groupby("session_id").\
apply(similarity).mean()

0.8356709936623352

In [ ]:
# implement session metric 1: time to first skip (number of songs to first skip)

In [ ]:
# implement session metric 2: sustained dissatisfaction: proportions of sessions with 3 consecutive skips

In [ ]:
# implement session metric 3: session coherence: average similarity between the top recommended tracks

Once these metrics are implemented, compute these metrics for topk=5 and topk=10 compare their estimates for the production ranker as a correlation plot. Please note which metrics are correlated with ndcg metric.

**Additional goal:**
Implement another simple ranking logic, and compare the performance of both the production ranker and new ranking policy on the ndcg and three behavioral metrics.
A simple ranking policy could include sortby track popularity, or sort by danceability score for the track.

Please report the performance of these rankers on all four metrics.

# Part B: Off Policy Evaluation

We log listener behavior based on the recommendations that the production recommender serves to the listener. Using this data to assess any new recommender system, however, can present challenges – the production recommender and the new recommender can drastically differ in the results that they display to the user. For example, maybe the new recommender presents a lot of niche content, while the production recommender presents a lot of popular options. This can be an issue when evaluating a new recommender – If you don’t have any feedback on a recommendation because you never presented it to a user, how can you evaluate whether it’s a good recommendation?
If you have a new policy to test that’s very similar to your old approach, then this won’t be an issue, and it’ll be easy to test! However, if the policy is very different, then you’ll need to collect special logged data.

In this part of the project, we will simulate a recommendation policy and leverage counterfactual estimators as metrics to compare performance.


Lets first begin by generating a few users and products. For ease of simulation, we assume users derive equal satisfaction from each item.

In [ ]:

users = np.array(["user1", "user2", "user3"])
products = np.array(
    [
        "product_a",
        "product_b",
        "product_c",
        "product_d",
        "product_e",
        "product_f",
        "product_g",
    ]
)

satisfaction = {
    "product_a": 100,
    "product_b": 150,
    "product_c": 100,
    "product_d": 200,
    "product_e": 500,
    "product_f": 120,
    "product_g": 160,
}



Lets also implement whether a given user will accept a given recommendation or not. Once done, we can implement a target policy that makes recommendations.

In [ ]:

def will_purchase(user, product):
    if user == "user1" and (
        product == "product_a" or product == "product_b" or product == "product_c"
    ):
        return True
    elif user == "user2" and (product == "product_d" or product == "product_e"):
        return True
    elif user == "user3" and (product == "product_f" or product == "product_g"):
        return True
    else:
        return False


def choose_user():
    return np.random.choice(users, size=1)


def logging_policy():
    return np.random.choice(products, size=1), 1 / len(products)


class TargetPolicy:
    def __init__(self):
        self.user_probs = {
            "user1": np.array([0.1, 0.1, 0.2, 0.1, 0.15, 0.15, 0.20]),
            "user2": np.array([0.1, 0.10, 0.05, 0.25, 0.3, 0.1, 0.1]),
            "user3": np.array([0.06, 0.06, 0.3, 0.06, 0.06, 0.4, 0.06]),
        }

        for user, probs in self.user_probs.items():
            assert probs.sum() == 1
            assert len(probs) == len(products)

    def recommend(self, user):
        user_prob = self.user_probs[user]
        product = np.random.choice(products, size=1, p=user_prob)
        product_idx = np.where(products == product)
        prob = user_prob[product_idx]

        return product, prob

    def get_prob(self, user, product):
        user_prob = self.user_probs[user]
        product_idx = np.where(products == product)
        product_prob = user_prob[product_idx]

        return product_prob



Having defined all key components of the dataset generation, lets create logged data that we can finally use for evaluation purposes:

In [ ]:
def compute_satisfaction(user, product):
    if will_purchase(user, product):
        return satisfaction[product.item()]
    else:
        return 0


def create_logs(n=1000):
    logs = []
    target_policy = TargetPolicy()

    for _ in range(n):
        user = choose_user()

        logging_product, logging_prob = logging_policy()
        model_prob = target_policy.get_prob(user.item(), logging_product)

        target_product, _ = target_policy.recommend(user.item())

        logging_satisfaction = compute_satisfaction(user, logging_product)
        target_satisfaction = compute_satisfaction(user, target_product)

        log = OrderedDict(
            {
                "user_features": user.item(),
                "item_placed": logging_product.item(),
                "item_prob": logging_prob,
                "item_satisfaction": logging_satisfaction,
                "model_prob": model_prob.item(),
                "ab_test_satisfaction": target_satisfaction,
            }
        )

        logs.append(log)

    return pd.DataFrame(logs)

Here is what ur logged data now looks like:

In [ ]:
logs = create_logs(n=1000)
logs.head(5)

user_features item_placed  item_prob  item_satisfaction  model_prob  \
0         user2   product_c   0.142857                  0        0.05   
1         user3   product_d   0.142857                  0        0.06   
2         user3   product_a   0.142857                  0        0.06   
3         user2   product_f   0.142857                  0        0.10   
4         user2   product_c   0.142857                  0        0.05   

   ab_test_satisfaction  
0                   500  
1                     0  
2                     0  
3                     0  
4                     0

In [ ]:
logs[["user_features", "item_placed", "item_prob", "item_satisfaction"]].sample(n=10)

user_features item_placed  item_prob  item_satisfaction
804         user3   product_c   0.142857                  0
603         user1   product_c   0.142857                100
772         user1   product_b   0.142857                150
279         user3   product_b   0.142857                  0
731         user1   product_g   0.142857                  0
698         user2   product_g   0.142857                  0
706         user2   product_g   0.142857                  0
987         user3   product_a   0.142857                  0
527         user1   product_b   0.142857                150
104         user2   product_b   0.142857                  0

With all the dataset ready, lets compute the mean rewards (satisfaction) for the logging/production policy and the target policy:

In [ ]:
%%time
sim = create_logs(n=100000)
logging_policy = sim["item_satisfaction"].mean()
target_policy = sim["ab_test_satisfaction"].mean()

print(f"Expected reward from logging policy: {logging_policy: .2f}")
print(f"Expected reward from target policy: {target_policy: .2f}")

Expected reward from logging policy:  63.36
Expected reward from target policy:  101.64
CPU times: user 6.98 s, sys: 7.91 ms, total: 6.99 s
Wall time: 6.99 s


Now lets implement the IPS estimator:

In [ ]:
def compute_ips(df):
    assert {"model_prob", "item_prob", "item_satisfaction"}.issubset(df.columns)
    return (df["model_prob"] / df["item_prob"] * df["item_satisfaction"]).mean()

In [ ]:
ips_est = compute_ips(logs)
ips_est

109.3414

Computing the IPS estimator on our 1,000 entry log gives an average revenue of 109.34 (very close to the true performance of 100.99) compared with the average revenue of the logging policy of 63.36. Therefore, we should be confident to deploy our target policy to production and do an AB test comparing it with the logging policy as a final validation.

## Goals for this part of project:

Finish the implementation of two additional off-policy estimators:
1. Capped IPS
2. Normalized Capped Importance Sampling (NCIS)

Feel free to try different capping thresholds, and compare the reward and standard deviations of these estimators with the IPS estimator and mean reward.


In [ ]:
def compute_capped_ips(logs, cap=1000):

    return 

In [ ]:
def compute_ncis(logs, cap=1000):

    return 